runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/Environments.java

```
@@ -207,31 +210,94 @@ public static Environment createProcessEnvironment(
     for (String path : pathsToStage) {
       File file = new File(path);
-      if (new File(path).exists()) {
-        // Spurious items get added to the classpath. Filter by just those that exist.
+      // Spurious items get added to the classpath. Filter by just those that exist.
+      if (file.exists()) {
+        ArtifactInformation.Builder artifactBuilder = ArtifactInformation.newBuilder();
+        artifactBuilder.setTypeUrn(BeamUrns.getUrn(StandardArtifacts.Types.FILE));
+        artifactBuilder.setRoleUrn(BeamUrns.getUrn(StandardArtifacts.Roles.STAGING_TO));
+        artifactBuilder.setRolePayload(
+            RunnerApi.ArtifactStagingToRolePayload.newBuilder()
+                .setStagedName(createStagingFileName(file))
+                .build()
+                .toByteString());
         if (file.isDirectory()) {
-          // Zip up directories so we can upload them to the artifact service.
+          File zippedFile;
+          HashCode hashCode;
           try {
-            filesToStage.add(createArtifactInformation(zipDirectory(file)));
+            zippedFile = zipDirectory(file);
+            hashCode = Files.asByteSource(zippedFile).hash(Hashing.sha256());
           } catch (IOException e) {
             throw new RuntimeException(e);
           }
+          artifactsBuilder.add(
+              artifactBuilder
+                  .setTypePayload(
+                      RunnerApi.ArtifactFilePayload.newBuilder()
+                          .setPath(zippedFile.getPath())
+                          .setSha256(hashCode.toString())
+                          .build()
+                          .toByteString())
+                  .build());
         } else {
-          filesToStage.add(createArtifactInformation(file));
+          HashCode hashCode;
+          try {
+            hashCode = Files.asByteSource(file).hash(Hashing.sha256());
+          } catch (IOException e) {
+            throw new RuntimeException(e);
```

Or would it be better to let the method throw an IOException?